src/macros.rs

```diff
@@ -444,3 +451,39 @@ macro_rules! valkey_module {
         }
     }
 }
+
+#[macro_export]
+macro_rules! valkey_module_auth {
+    ($ctx:expr, $($auth_callback:expr),* $(,)*) => {
+        $(
+            {   // New scope for each callback
+                extern "C" fn __do_auth(
+                    ctx: *mut $crate::raw::RedisModuleCtx,
+                    username: *mut $crate::raw::RedisModuleString,
+                    password: *mut $crate::raw::RedisModuleString,
+                    err: *mut *mut $crate::raw::RedisModuleString,
+                ) -> std::os::raw::c_int {
+                    let context = $crate::Context::new(ctx);
+                    let ctx_ptr = unsafe { std::ptr::NonNull::new_unchecked(ctx) };
+                    let username = $crate::ValkeyString::new(Some(ctx_ptr), username);
+                    let password = $crate::ValkeyString::new(Some(ctx_ptr), password);
+
+                    match $auth_callback(&context, username, password) {
+                        Ok(result) => result,
+                        Err(e) => {
+                            let error_msg = $crate::ValkeyString::create(None, e.to_string().as_str());
+                            unsafe { *err = error_msg.into_raw() };
+                            $crate::AUTH_HANDLED
+                        }
+                    }
+                }
+
+                unsafe {
+                    if let Some(register_auth) = $crate::raw::RedisModule_RegisterAuthCallback {
```

Comment:
1. Lets make this code `expect` that the RedisModule_RegisterAuthCallback API exists.

```suggestion
                    $crate::raw::RedisModule_RegisterAuthCallback.expect("ValkeyModule_RegisterAuthCallback should exist on engines 7.2 and above")($ctx, Some(__do_auth));
```

2. We should be using feature flags here to ensure this code does not work on engine versions before 7.2